File: lib/features/player/models/filter_preset.dart
import 'package:clashkingapp/features/player/models/war_stats_filter.dart';

class FilterPreset {
  final String id;
  final String name;
  final WarStatsFilter filter;
  final DateTime createdAt;

  const FilterPreset({
    required this.id,
    required this.name,
    required this.filter,
    required this.createdAt,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'filter': filter.toJson(),
      'createdAt': createdAt.millisecondsSinceEpoch,
    };
  }

  factory FilterPreset.fromJson(Map<String, dynamic> json) {
    return FilterPreset(
      id: json['id'],
      name: json['name'],
      filter: WarStatsFilter.fromJson(json['filter']),
      createdAt: DateTime.fromMillisecondsSinceEpoch(json['createdAt']),
    );
  }

  FilterPreset copyWith({
    String? id,
    String? name,
    WarStatsFilter? filter,
    DateTime? createdAt,
  }) {
    return FilterPreset(
      id: id ?? this.id,
      name: name ?? this.name,
      filter: filter ?? this.filter,
      createdAt: createdAt ?? this.createdAt,
    );
  }
}

File: expo/src/features/player/models/filter-preset.ts
import { WarStatsFilter } from './war-stats-filter';
import { int, record, string, type JsonRecord } from './parsing';
export class FilterPreset {
  constructor(
    readonly id: string,
    readonly name: string,
    readonly filter: WarStatsFilter,
    readonly createdAt: Date,
  ) {}
  toJson(): JsonRecord {
    return {
      id: this.id,
      name: this.name,
      filter: this.filter.toJson(),
      createdAt: this.createdAt.getTime(),
    };
  }
  static fromJson(json: JsonRecord) {
    return new FilterPreset(
      string(json.id),
      string(json.name),
      WarStatsFilter.fromJson(record(json.filter)),
      json.createdAt == null ? new Date() : new Date(int(json.createdAt)),
    );
  }
  copyWith(value: Partial<{ id: string; name: string; filter: WarStatsFilter; createdAt: Date }>) {
    return new FilterPreset(
      value.id ?? this.id,
      value.name ?? this.name,
      value.filter ?? this.filter,
      value.createdAt ?? this.createdAt,
    );
  }
}
